packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -688,154 +861,585 @@ class _SortData {
+      // ancestor for all nodes. This provides a base directionality to use for
+      // the ordering of the groups.
+      final TextDirection nearestCommonDirectionality = _ReadingOrderSortData.commonDirectionalityOf(inBandOfTop);
+
+      // Do an initial common-directionality-based sort to get consistent geometric
+      // ordering for grouping into directionality groups. It has to use the
+      // common directionality to be able to group into sane groups for the
+      // given directionality, since rectangles can overlap and give different
+      // results for different directionalities.
+      mergeSort<_ReadingOrderSortData>(inBandOfTop, compare: (_ReadingOrderSortData a,_ReadingOrderSortData b) {
+        switch(nearestCommonDirectionality) {
+          case TextDirection.ltr:
+            return a.rect.left.compareTo(b.rect.left);
+          case TextDirection.rtl:
+            return b.rect.right.compareTo(a.rect.right);
+        }
+        assert(false, 'Unhandled directionality $nearestCommonDirectionality');
+        return 0;
+      });
+
+      // Collect the top band into internally sorted groups with shared directionality.
+      final List<_ReadingOrderDirectionalGroupData> bandGroups = collectDirectionalityGroups(inBandOfTop);
+      if (bandGroups.length == 1) {
+        // There's only one directionality group, so just send back the first
+        // one in that group, since it's already sorted.
+        return bandGroups.first.members.first;
       }
-      return topmost;
+
+      // Sort the groups based on the common directionality and bounding boxes.
+      mergeSort<_ReadingOrderDirectionalGroupData>(bandGroups, compare: (_ReadingOrderDirectionalGroupData a, _ReadingOrderDirectionalGroupData b) {
+        switch(nearestCommonDirectionality) {
+          case TextDirection.ltr:
+            return a.rect.left.compareTo(b.rect.left);
+          case TextDirection.rtl:
+            return b.rect.right.compareTo(a.rect.right);
+        }
+        assert(false, 'Unhandled directionality $nearestCommonDirectionality');
+        return 0;
+      });
+      return bandGroups.first.members.first;
     }
 
-    final List<_SortData> data = <_SortData>[
-      for (final FocusNode node in nodes) _SortData(node),
+    final List<_ReadingOrderSortData> data = <_ReadingOrderSortData>[
+      for (final FocusNode node in descendants) _ReadingOrderSortData(node),
     ];
 
     // Pick the initial widget as the one that is leftmost in the band of the
     // topmost, or the topmost, if there are no others in its band.
-    final List<_SortData> sortedList = <_SortData>[];
-    final List<_SortData> unplaced = data.toList();
-    _SortData current = pickFirst(unplaced);
+    final List<_ReadingOrderSortData> sortedList = <_ReadingOrderSortData>[];
+    final List<_ReadingOrderSortData> unplaced = data.toList();
+    _ReadingOrderSortData current = pickNext(unplaced);
```

Comment:
How performant is it actually to do all this `pickNext` work for every single node? Given a Desktop screen with lots of elements, is this actually fast enough to figure out the next focus when tap is pressed?

Reply:
It's OK as is: I don't notice any delay on a phone or a desktop. I'd like it faster, of course, and I tried to make an algorithm that would just sort things in one pass, but it needs to ignore things that have already been placed each time it picks the next one, since it has to consider the "band" that the first one is in.  I think it's probably still possible to consolidate it into one pass, but it'll make it even more complex than it already is.  I'll take another look at doing that, but maybe in another PR?